[GPU Process] Images with certain colorspaces cannot be rendered

shallawa opened the pull request:
#### 9d23c40ee7ce3c80b6d2b40cc534e78cac079019
<pre>
[GPU Process] Images with certain colorspaces cannot be rendered
<a href="https://bugs.webkit.org/show_bug.cgi?id=266300">https://bugs.webkit.org/show_bug.cgi?id=266300</a>
<a href="https://rdar.apple.com/118238178">rdar://118238178</a>

Reviewed by Cameron McCormack.

Check if the color space of the NativeImage is supported by CG for output before
creating ShareableBitmap. If it is not supported, fall back to SRGB.

* LayoutTests/fast/images/resources/green-100x100-unsupported-output-color-space.jpg: Added.
* LayoutTests/fast/images/unsupported-output-color-space-expected.html: Added.
* LayoutTests/fast/images/unsupported-output-color-space.html: Added.
* Source/WebKit/Shared/ShareableBitmap.cpp:
(WebKit::ShareableBitmap::createFromImageDraw):
* Source/WebKit/Shared/ShareableBitmap.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.cpp:
(WebKit::createShareableBitmapFromNativeImage)

Canonical link: <a href="https://commits.webkit.org/273204@main">https://commits.webkit.org/273204@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/56975ee54ac6506f55f78e5ca93794fc418f849a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/34655 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/13476 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/36661 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/37338 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/31299 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/15880 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/10575 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/30261 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/35178 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/11407 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/30876 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/9966 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/10047 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/30950 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/38613 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/31467 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/31263 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/36100 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/10167 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/8052 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/34060 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/11988 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/30575 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7957 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/10715 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/11041 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

heycam commented on Source/WebKit/Shared/ShareableBitmap.cpp:
```diff
@@ -109,8 +109,17 @@ RefPtr<ShareableBitmap> ShareableBitmap::createFromImageDraw(NativeImage& image)
+    if (!context) {
+        // If createGraphicsContext() failed because the image colorSpace is not
+        // supported for output, fallback to SRGB.
+        bitmap = ShareableBitmap::create({ imageSize, DestinationColorSpace::SRGB() });
```
Not sure if it's a big deal, but should we release the old `ShareableBitmap` before creating this new one, to avoid memory being allocated for both at the same time? Or, maybe we should move the fallback into `ShareableBitmap` somehow so it can re-use the same allocation?

shallawa committed:
[GPU Process] Images with certain colorspaces cannot be rendered
https://bugs.webkit.org/show_bug.cgi?id=266300
rdar://118238178

Reviewed by Cameron McCormack.

Check if the color space of the NativeImage is supported by CG for output before
creating ShareableBitmap. If it is not supported, fall back to SRGB.

* LayoutTests/fast/images/resources/green-100x100-unsupported-output-color-space.jpg: Added.
* LayoutTests/fast/images/unsupported-output-color-space-expected.html: Added.
* LayoutTests/fast/images/unsupported-output-color-space.html: Added.
* Source/WebKit/Shared/ShareableBitmap.cpp:
(WebKit::ShareableBitmap::createFromImageDraw):
* Source/WebKit/Shared/ShareableBitmap.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.cpp:
(WebKit::createShareableBitmapFromNativeImage)

Canonical link: https://commits.webkit.org/273204@main
